### PR TITLE
[SPARK-11178] Improving naming around task failures.

### PR DESCRIPTION
**Pull request by kayousterhout**
Commit af3bc59d1f5d9d952c2d7ad1af599c49f1dbdaf0 introduced new
functionality so that if an executor dies for a reason that's not
caused by one of the tasks running on the executor (e.g., due to
pre-emption), Spark doesn't count the failure towards the maximum
number of failures for the task.  That commit introduced some vague
naming that this commit attempts to fix; in particular:

(1) The variable "isNormalExit", which was used to refer to cases where
the executor died for a reason unrelated to the tasks running on the
machine, has been renamed (and reversed) to "exitCausedByApp". The problem
with the existing name is that it's not clear (at least to me!) what it
means for an exit to be "normal"; the new name is intended to make the
purpose of this variable more clear.

(2) The variable "shouldEventuallyFailJob" has been renamed to
"countTowardsTaskFailures". This variable is used to determine whether
a task's failure should be counted towards the maximum number of failures
allowed for a task before the associated Stage is aborted. The problem
with the existing name is that it can be confused with implying that
the task's failure should immediately cause the stage to fail because it
is somehow fatal (this is the case for a fetch failure, for example: if
a task fails because of a fetch failure, there's no point in retrying,
and the whole stage should be failed).

### CODE REVIEW

**Comment by mccheah** [comment]
Thanks for picking up some things we missed!

"In YarnAllocator.scala, line 447: what does "exited normally" mean here? I'm hoping to improve the comment to be clearer to those less familiar with YARN." - "exited normally" is a case when an executor would exit not because of the user's code or the job itself, but because of events that occur during the YARN application lifecycle. In other words, the executor does not exit because of problems, per se, but exits in "normal" operation. The definition of "normal" of course is still vague.

"In SparkDeploySchedulerBackend, the original commit seems to have changed the behavior..." - this might unfortunately be a bug in how I refactored things. How would this manifest? @kayousterhout , would it be possible for you to reproduce it and fix if necessary?

Changing JsonProtocol seems fine.

"For the case where the commit failed, we never count that towards the max failures. Is there any legimate reason the commit could fail? I'm wondering if it's possible to have a stage that fails infinitely many times due to a real problem with committing the task." - the TaskCommitDenied failed event is explicitly coordinated by the driver, and is fired only if the driver saw that a previous copy of this task had already started committing the output before this one. This is relevant in speculative execution mode. TaskCommitDenied isn't fired if the commit itself fails however. See SparkHadoopMapRedUtil.commitTask.

**Comment by vanzin** [comment]
> (a) In YarnAllocator.scala, line 447: what does "exited normally"

That means the container exited with exit status 0.

> (c) This commit changes the JsonProtocol. I think this is OK because the existing Json was never part of a release

Correct.

> (d) I don't love using a negative in the variable name ("exitUnrelatedToRunningTasks")

I don't love that name either. Maybe `exitCausedByApp`? `exitIsAppRelated`?

> (e) Is there any legimate reason the commit could fail?

Intermittent HDFS or network issues, for example.

**Comment by mccheah** [comment]
@vanzin - see my earlier comment on the commit failing case. This particular event that is handled is in the case that multiple copies of the same file-writing task try to commit output.

**Comment by vanzin** [comment]
@mccheah ah you're right; I assumed she was asking about when commits would fail in general.

**Comment by kayousterhout** [comment]
Just filed #9273 about the bug and updated the comments / naming based on @vanzin's recommendation

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorLossReason.scala [hunk 1]
```diff
@@ -28,12 +28,15 @@ class ExecutorLossReason(val message: String) extends Serializable {
...
 
 private[spark]
-case class ExecutorExited(exitCode: Int, isNormalExit: Boolean, reason: String)
+case class ExecutorExited(exitCode: Int, exitCausedByApp: Boolean, reason: String)
   extends ExecutorLossReason(reason)
 
 private[spark] object ExecutorExited {
-  def apply(exitCode: Int, isNormalExit: Boolean): ExecutorExited = {
-    ExecutorExited(exitCode, isNormalExit, ExecutorExitCode.explainExitCode(exitCode))
+  def apply(exitCode: Int, exitUnrelatedToRunningTasks: Boolean): ExecutorExited = {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/ExecutorLossReason.scala]: Did you forget to update this?

**Comment by vanzin** [comment]
Seems like you have to rebase; could you also update the description if the questions have been answered?

**Comment by vanzin** [comment]
LGTM.

**Comment by kayousterhout** [comment]
Jenkins, retest this please

**Comment by squito** [comment]
lgtm

**Pull request closed without merging**